agent/agent.go

```diff
@@ -206,6 +206,7 @@ func (a *Agent) run(ctx context.Context) {
 		leaving       = a.leaving
 		subscriptions = map[string]context.CancelFunc{}
 	)
+	defer session.close()
```

Comment:
Careful here. I believe this would call the `close` method on the original `session`.

> Each time a "defer" statement executes, the function value and parameters to the call are evaluated as usual and saved anew but the actual function is not invoked.

https://golang.org/ref/spec#Defer_statements